### DISCUSSION TITLE
[6 Luglio 2020] - Pipeline

### DISCUSSION THREAD

**Discussion by notedo** [CS-Swap/Architettura-degli-Elaboratori#26] [Pipeline]
### Traccia Esercizio

Si consideri l'architettura MIPS con pipeline mostrata nella figura in basso. Il programma qui di seguito effettua le operazioni di somma o sottrazione indicate nella stringa fornita in input.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
1) le istruzioni tra le quali sono presenti data hazard – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
2) le istruzioni tra le quali sono presenti control hazard – indicare i numeri di istruzione N ed M (nel formato N / M);
3) quanti cicli di clock sono necessari ad eseguire il programma tramite forwarding, spiegando il calcolo effettuato;
4) quanti cicli di clock sarebbero necessari ad eseguire il programma senza forwarding, spiegando il calcolo effettuato;
5) quali sono, per ognuna delle cinque fasi, le istruzioni (o le bolle) in pipeline durante l’11° ciclo di clock (con forwarding);

<img width="546" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/f12e2306-9957-4fb6-9bfe-fec8341aa6c8">

<img width="870" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/6f19552e-7573-42c1-9724-365a0f76117c">

### Soluzione prof
[AE2.2019-20.-.Appello.2.Esercizio.4.-.2020-07-06.corretto.pdf](https://github.com/user-attachments/files/15518347/AE2.2019-20.-.Appello.2.Esercizio.4.-.2020-07-06.corretto.pdf)

**Comment by thegeek-sys** [comment]
# Senza forwarding
<img width="958" alt="Screenshot 2024-06-03 alle 22 57 25" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/4df01e5a-a52e-43d5-a982-8c55ae60a3a9">

# Con forwarding
<img width="958" alt="Screenshot 2024-06-03 alle 22 58 34" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/4f9ef1da-fc4b-43e7-9025-7f1c8df7a0b8">

# Data hazard e control hazard
<img width="248" alt="Screenshot 2024-06-03 alle 22 58 47" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/e82de6e7-1ef5-4d34-b07a-a4feacae086c">

# Istruzioni 11° ciclo di clock
<img width="86" alt="Screenshot 2024-06-03 alle 22 58 54" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/519593d6-b1ac-46e5-a273-0a81caa27fa6">
